# Global Register of Introduced and Invasive Species - Australia (GRIIS)
Import GRIIS list from [GBIF IPT](https://cloud.gbif.org/griis/resource?r=griis-australia)

In [139]:
#import essential libraries
import pandas as pd
import requests
import os
import zipfile
from pathlib import Path

Download the Darwin Core Archive and read the species list into a dataframe

In [140]:
url = "https://cloud.gbif.org/griis/archive.do?r=griis-australia&v=1.6"
basedir = "/Users/new330/IdeaProjects/authoritative-lists/"
sourcedir = basedir+"source-data/griis/"
targetdir = basedir+"current-lists/griis/"
print("Downloading dwca")
r = requests.get(url)

Retrieve the filename from the headers. Download and unzip the file

In [144]:
#get the filename from the headers and create a directory holder for it
basedir = "/Users/new330/IdeaProjects/authoritative-lists/"
sourcedir = basedir+"source-data/griis/"
targetdir = basedir+"current-lists/griis/"
dwca = sourcedir + r.headers['content-disposition'].split("=")[1].replace('"','')
if Path(dwca).exists():
  os.remove(dwca)
unzipdirname = dwca[:-len(".zip")]
if not Path(unzipdirname):
  os.mkdir(unzipdirname)
#downlaad
with open(dwca, "wb") as output_file:
    output_file.write(r.content)
# unzip
with zipfile.ZipFile(dwca, 'r') as z:
   z.extractall(unzipdirname)

Explore the list, looking especially for encoding issues.

In [146]:
#explore
taxondf = pd.read_csv(unzipdirname + os.sep + "taxon.txt",sep="\t",lineterminator="\n")
taxondf

id  taxonID                                   scientificName  \
0     151000   151000  Abelia ×grandiflora (Rovelli ex André) Rehder   
1     151001   151001                  Abelmoschus manihot (L.) Medik.   
2     151002   151002                      Abutilon grandiflorum G.Don   
3     151003   151003         Abutilon pictum (Gillies ex Hook.) Walp.   
4     151004   151004                      Abutilon theophrasti Medik.   
...      ...      ...                                              ...   
2979  153979   153979                            Ziziphus jujuba Mill.   
2980  153980   153980                         Ziziphus mauritiana Lam.   
2981  153981   153981                        Ziziphus mucronata Willd.   
2982  153982   153982                Ziziphus spina-christi (L.) Desf.   
2983  153983   153983    Zoobotryon verticillatum (Delle Chiaje, 1822)   

                                      acceptedNameUsage   kingdom  \
0                                                   NaN   Plantae   
1                                                   NaN   Plantae   
2                                                   NaN   Plantae   
3     Callianthe picta (Gillies ex Hook. & Arn.) Don...   Plantae   
4                                                   NaN   Plantae   
...                                                 ...       ...   
2979                                                NaN   Plantae   
2980                                                NaN   Plantae   
2981                                                NaN   Plantae   
2982                                                NaN   Plantae   
2983          Amathia verticillata (delle Chiaje, 1822)  Animalia   

            phylum          class           order          family taxonRank  \
0     Tracheophyta  Magnoliopsida      Dipsacales  Caprifoliaceae   SPECIES   
1     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
2     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
3     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
4     Tracheophyta  Magnoliopsida        Malvales       Malvaceae   SPECIES   
...            ...            ...             ...             ...       ...   
2979  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2980  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2981  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2982  Tracheophyta  Magnoliopsida         Rosales      Rhamnaceae   SPECIES   
2983       Bryozoa   Gymnolaemata  Ctenostomatida  Vesiculariidae   SPECIES   

     taxonomicStatus  
0           ACCEPTED  
1           ACCEPTED  
2           ACCEPTED  
3            SYNONYM  
4           ACCEPTED  
...              ...  
2979        ACCEPTED  
2980        ACCEPTED  
2981        ACCEPTED  
2982        ACCEPTED  
2983         SYNONYM  

[2984 rows x 11 columns]

Problem records found by eyeballing. The file specifies UTF-8 encoding but has been encoded in something else.

In [147]:
taxondf.loc[taxondf['taxonID'].isin([151000,151155,151156,151617,151634,152066,153106])][['taxonID','scientificName']]

taxonID                                   scientificName
0      151000  Abelia ×grandiflora (Rovelli ex André) Rehder
155    151155         Amphilophus citrinellus (Günther, 1864)
156    151156            Amphilophus labiatus (Günther, 1864)
617    151617         Cichlasoma trimaculatum (Günther, 1867)
634    151634               Cladophora prolifera (Roth) Kütz.
1066   152066               Fallopia convolvulus (L.) Á.Löve
2106   153106                      Pinus durangensis Martínez

These values should be:

151000 Abelia x grandiflora (Rovelli ex André) Rehder
151155 Amphilophus citrinellus (Günther, 1864)
151156 Amphilophus labiatus (Günther, 1864)
153106 Pinus durangensis Martínez

There are many more though, just by searching for a single character:

In [148]:
taxondf.loc[taxondf['scientificName'].str.contains('Ã')]

id  taxonID                                     scientificName  \
0     151000   151000    Abelia ×grandiflora (Rovelli ex André) Rehder   
104   151104   151104                     Aloe parvibracteata Schönland   
122   151122   151122     Alysicarpus ovalifolius (Schumach.) J.Léonard   
148   151148   151148               Amniataba percoides (Günther, 1864)   
155   151155   151155           Amphilophus citrinellus (Günther, 1864)   
156   151156   151156              Amphilophus labiatus (Günther, 1864)   
195   151195   151195          Antithamnion cruciatum (C.Agardh) Nägeli   
213   151213   151213       Archocentrus nigrofasciatus (Günther, 1867)   
249   151249   151249                 Ascidiella aspersa (Müller, 1776)   
265   151265   151265             Astatotilapia burtoni (Günther, 1894)   
450   151450   151450                Campsis ×tagliabuana (Vis.) Rehder   
509   151509   151509                           Castilla elastica Sessé   
617   151617   151617           Cichlasoma trimaculatum (Günther, 1867)   
634   151634   151634                 Cladophora prolifera (Roth) Kütz.   
696   151696   151696     Cortaderia jubata (Lemoine ex Carrière) Stapf   
712   151712   151712                   Cottoniella fusiformis Børgesen   
766   151766   151766                              Croton hirtus L'Hér.   
862   151862   151862                   Dianella ensifolia (L.) Redouté   
898   151898   151898                Dipolydora flava (Claparède, 1870)   
1015  152015   152015                    Erodium cicutarium (L.) L'Hér.   
1016  152016   152016                    Erodium malacoides (L.) L'Hér.   
1017  152017   152017                Erodium moschatum (Burm.f.) L'Hér.   
1066  152066   152066                 Fallopia convolvulus (L.) Á.Löve   
1161  152161   152161                Geranium yeoi C.Aedo & Muñoz Garm.   
1260  152260   152260      Hemidactylus frenatus Duméril & Bibron, 1836   
1270  152270   152270           Heterotis rotundifolia (Sm.) Jacq.-Fél.   
1271  152271   152271  Hevea brasiliensis (Willd. ex A.Juss.) Müll.Arg.   
1307  152307   152307              Hydroides sanctaecrucis Krøyer, 1863   
1329  152329   152329                      Hypericum kouytchense H.Lév.   
1330  152330   152330                        Hypericum moserianum André   
1518  152518   152518  Lepidodactylus lugubris (Duméril & Bibron, 1836)   
1584  152584   152584                        Lonicera ×americana K.Koch   
1620  152620   152620                Lygosoma bowringii (Günther, 1864)   
1656  152656   152656                  Manihot carthaginensis Müll.Arg.   
1854  152854   152854          Oncosiphon piluliferum (L.f.) Källersjö   
1855  152855   152855          Oncosiphon suffruticosum (L.) Källersjö   
1927  152927   152927                      Oxalis valdiviensis Barnéoud   
1978  152978   152978           Paspalum fasciculatum Willd. ex Flüggé   
2016  153016   153016                   Pelargonium quercifolium L'Hér.   
2035  153035   153035                 Petasites pyrenaicus (L.) G.López   
2037  153037   153037           Petrorhagia dubia (Raf.) G.López & Romo   
2106  153106   153106                        Pinus durangensis Martínez   
2148  153148   153148                    Plectranthus fruticosus L'Hér.   
2281  153281   153281                         Quercus palustris Münchh.   
2355  153355   153355            Rosmarinus officinalis de Noë ex Lange   
2472  153472   153472      Schoenoplectus californicus (C.A.Mey.) Soják   
2481  153481   153481    Scleropages formosus (Müller & Schlegel, 1840)   
2538  153538   153538                Setaria palmifolia (J.König) Stapf   
2539  153539   153539              Setaria parviflora (Poir.) Kerguélen   
2614  153614   153614        Solieria filiformis (Kütz.) P.W.Gabrielson   
2681  153681   153681                   Stuckenia pectinata (L.) Börner   
2924  153924   153924  Vigna adenantha (E.Mey.) Maréchal,

ftfy (fixed that for you) library works magic on these

In [149]:
import ftfy
from ftfy import fix_encoding

sample = "Abelia ×grandiflora (Rovelli ex André) Rehder"
fix_encoding(sample)

'Abelia ×grandiflora (Rovelli ex André) Rehder'

Apply to the `scientificName` and `acceptedNameUsage` fields

In [150]:
taxondf['scientificName'] = taxondf['scientificName'].apply(fix_encoding)
taxondf[['taxonID','scientificName','acceptedNameUsage']]

taxonID                                 scientificName  \
0      151000  Abelia ×grandiflora (Rovelli ex André) Rehder   
1      151001                Abelmoschus manihot (L.) Medik.   
2      151002                    Abutilon grandiflorum G.Don   
3      151003       Abutilon pictum (Gillies ex Hook.) Walp.   
4      151004                    Abutilon theophrasti Medik.   
...       ...                                            ...   
2979   153979                          Ziziphus jujuba Mill.   
2980   153980                       Ziziphus mauritiana Lam.   
2981   153981                      Ziziphus mucronata Willd.   
2982   153982              Ziziphus spina-christi (L.) Desf.   
2983   153983  Zoobotryon verticillatum (Delle Chiaje, 1822)   

                                      acceptedNameUsage  
0                                                   NaN  
1                                                   NaN  
2                                                   NaN  
3     Callianthe picta (Gillies ex Hook. & Arn.) Don...  
4                                                   NaN  
...                                                 ...  
2979                                                NaN  
2980                                                NaN  
2981                                                NaN  
2982                                                NaN  
2983          Amathia verticillata (delle Chiaje, 1822)  

[2984 rows x 3 columns]

Write the file and remove the original zip download.

In [151]:
taxondf.to_csv(targetdir+"griis-1.6.csv",index=False)
os.remove(dwca)